lib/awful/screenshot.lua

```diff
@@ -0,0 +1,884 @@
+      return d
+    end
+  end
+
+  return nil
+end
+
+-- Internal function to get the default filename prefix
+local function get_default_prefix()
+  return "Screenshot-"
+end
+
+-- Internal function to check a directory string for existence and writability.
+-- This only checks if the requested directory exists and is writeable, not if
+-- such a directory is legal (i.e. it behaves as 'mkdir' and not 'mkdir -p').
+-- Adding 'mkdir -p' functionality can be considered in the future.
+local function check_directory(directory)
+
+  if directory and type(directory) == "string"  then
+
+    -- Fully qualify a "~/" path or a relative path to $HOME. One might argue
+    -- that the relative path should fully qualify to $PWD, but for a window
+    -- manager this should be the same thing to the extent that anything else
+    -- is arguably unexpected behavior.
+    if string.find(directory, "^~/") then
+      directory = string.gsub(directory, "^~/",
+                              string.gsub(os.getenv("HOME"), "/*$", "/"))
+    elseif string.find(directory, "^[^/]") then
+      directory = string.gsub(os.getenv("HOME"), "/*$", "/") .. directory
+    end
+
+    -- Assure that we return exactly one trailing slash
+    directory = string.gsub(directory, '/*$', '/')
+
+    -- If we use single quotes, we only need to deal with single quotes - (I
+    -- promise that's meaningful if you think about it from a bash perspective)
+    local dir = string.gsub(directory, "'", "'\\'\\\\\\'\\''")
+
+    if os.execute("bash -c 'if [ -d '\\''" .. dir .. "'\\'' -a -w '\\''" ..
+                  dir .. "'\\'' ] ; then exit 0 ; else exit 1 ; fi'") then
+      return directory
+    else
+      -- Currently returns nil if the requested directory string cannot be used.
+      -- This can be swapped to a silent fallback to the default directory if
+      -- desired. It is debatable which way is better.
+      return nil
+    end
+
+  else
+    -- No directory argument means use the default. Technically an outrageously
+    -- invalid argument (i.e. not even a string) currently falls back to the
+    -- default as well.
+    return get_default_dir()
+  end
+
+end
+
+-- Internal function to sanitize a prefix string
+--
+-- Currently only strips all path separators ('/'). Allows for empty prefix.
+local function check_prefix(prefix)
+  -- Maybe add more sanitizing eventually
+  if prefix and type(prefix) == "string" then
+    prefix = string.gsub(prefix, '/', '')
+    return prefix
+  end
+  return get_default_prefix()
+end
+
+-- Internal routine to verify that a filepath is valid.
+local function check_filepath(filepath)
+
+  -- This module is forcing png for now. In the event of adding more
+  -- options, this function is basically unchanged, except for trying
+  -- to match a regex for each supported format (e.g. (png|jpe?g|gif|bmp))
+  -- NOTE: we should maybe make this case insensitive too?
+  local fname_start, fname_end = string.find(filepath,'/[^/]+%.png$')
```

Comment:
is it start of file *name* or file path?

Reply:
Name. The regex is finding the last '/' through a path ending in '.png', so that parses a valid file name from a full path.